cirq/linalg/transformations.py

```diff
@@ -304,3 +317,152 @@ def partial_trace(tensor: np.ndarray,
+    `wavefunction` is expressed over N qubits. Attempts to factor the
+    wavefunction into a pure state over `keep_indices`. If this fails, a non-
+    unique mixture can be computed by eigendecomposition on the partial trace of
+    the input's density matrix.
+
+    States in the output mixture will retain the same type of shape as the input
+    wavefunction, either `(2 ** k)` or `(2, 2, ..., 2)` where k is the number of
+    qubits kept.
+
+    Args:
+        wavefunction: A wavefunction to express over a qubit subset.
+        keep_indices: Which indices to express the wavefunction on.
+        atol: The tolerance for determining that a factored state is pure.
+
+    Returns:
+        A single-component mixture in which the factored wavefunction has
+        probability '1' if the factored state is pure, or else a mixture of the
+        default eigendecomposition of the mixed state's partial trace.
+
+    Raises:
+        ValueError: if the input wavefunction shape is neither `(2 ** N)` nor
+        `(2, 2, ..., 2)`
+    """
+
+    # Attempt to do efficient state factoring.]
+    state = subwavefunction(wavefunction, keep_indices, default=None, atol=atol)
+    if state is not None:
+        return ((1.0, state),)
+
+    # Fall back to a (non-unique) mixture representation.
+    if wavefunction.shape == (wavefunction.size,):
+        return_shape = -1
+    elif wavefunction.shape == (2,) * int(np.log2(wavefunction.size)):
+        return_shape = (2,) * len(keep_indices)
+
+    keep_dims = 1 << len(keep_indices)
+    tot_dims = 1 << len(wavefunction.shape)
+    rho = np.kron(np.conj(wavefunction.reshape(tot_dims, 1)).T,
+                  wavefunction.reshape(tot_dims, 1)).reshape(
+                  (2, 2) * len(wavefunction.shape))
+    keep_rho = partial_trace(rho, keep_indices).reshape((keep_dims,) * 2)
+    eigvals, eigvecs = np.linalg.eig(keep_rho)
+    mixture = tuple(
+        zip(eigvals, [vec.reshape(return_shape) for vec in eigvecs.T]))
+    return [p for p in mixture if not approx_eq(p[0], 0.0)]
+
+
+def subwavefunction(wavefunction: np.ndarray,
+                    keep_indices: List[int],
+                    *,
+                    default: TDefault = RaiseValueErrorIfNotProvided,
+                    atol: Union[int, float] = 1e-8) -> np.ndarray:
+    r"""Return a representation of the given state over a subset of its qubits.
+
+    The input wavefunction must have shape `(2, 2, ..., 2)` or `(2 ** N)` where
+    `wavefunction` is expressed over N qubits. The returned array will retain
+    the same type of shape as the input wavefunction, either `(2 ** k)` or
+    `(2, 2, ..., 2)` where k is the number of qubits kept.
+
+    If a wavefunction |\psi> defined on N qubits is an outer product of kets
+    like  |\psi> = |x> \otimes |y>, and |x> is defined over the subset
+    `keep_indices` of  N qubits, then this method will factor |\psi> into |x>
+    and |y> and return |x>.  Global phase of the input state will not
```

Comment:
I would rephrase as

```
Note that $|x\rangle$ is not unique, because
$(e^{i \theta} |y\rangle) \otimes (|x\rangle) = (|y\rangle) \otimes (e^{i \theta} |x\rangle)$
. This method randomizes the global phase of $|x\rangle$ in order to avoid
accidental reliance on it.
```